throttle/contract_tx_limiter_middleware.go

```diff
@@ -0,0 +1,166 @@
+	if int64(contractInfo.TierInfo.MaxTx) > blockTxn.txn {
+		return false
+	} else {
+		return true
+	}
+}
+
+func (txl *contractTxLimiter) updateState(contractAddr loom.Address, curBlockHeight int64) {
+	blockTx, ok := txl.contractToBlockTrx[contractAddr.String()]
+	if !ok {
+		txl.contractToBlockTrx[contractAddr.String()] = blockTxn{1, curBlockHeight}
+		return
+	}
+	// reset blockTxn if block is out of range
+	if blockTx.blockHeight < curBlockHeight-
+		int64(txl.contractToInfoMap[contractAddr.String()].TierInfo.BlockRange) {
+		txl.contractToBlockTrx[contractAddr.String()] = blockTxn{0, curBlockHeight}
+		return
+	}
+	blockTx.txn++
+	txl.contractToBlockTrx[contractAddr.String()] = blockTx
+}
+
+// NewContractTxLimiterMiddleware add another tx limiter that limits how many CallTx(s) can be sent to an EVM contract within a pre-configured block range
+func NewContractTxLimiterMiddleware(cfg *ContractTxLimiterConfig,
+	createUserDeployerWhitelistCtx func(state loomchain.State) (contractpb.Context, error),
+) loomchain.TxMiddlewareFunc {
+	TxLimiter = &contractTxLimiter{
+		contractToBlockTrx: make(map[string]blockTxn, 0),
+	}
+	return loomchain.TxMiddlewareFunc(func(
+		state loomchain.State,
+		txBytes []byte,
+		next loomchain.TxHandlerFunc,
+		isCheckTx bool,
+	) (res loomchain.TxHandlerResult, err error) {
+		if !isCheckTx {
+			return next(state, txBytes, isCheckTx)
+		}
+		var nonceTx auth.NonceTx
+		if err := proto.Unmarshal(txBytes, &nonceTx); err != nil {
+			return res, errors.Wrap(err, "throttle: unwrap nonce Tx")
+		}
+		var tx loomchain.Transaction
+		if err := proto.Unmarshal(nonceTx.Inner, &tx); err != nil {
+			return res, errors.New("throttle: unmarshal tx")
+		}
+		if tx.Id != callId {
+			return next(state, txBytes, isCheckTx)
+		}
+		var msg vm.MessageTx
+		if err := proto.Unmarshal(tx.Data, &msg); err != nil {
+			return res, errors.Wrapf(err, "unmarshal message tx %v", tx.Data)
+		}
+		var msgTx vm.CallTx
+		if err := proto.Unmarshal(msg.Data, &msgTx); err != nil {
+			return res, errors.Wrapf(err, "unmarshal call tx %v", msg.Data)
+		}
+		if msgTx.VmType == vm.VMType_EVM {
+			if TxLimiter.contractToInfoMap == nil {
+				ctx, err := createUserDeployerWhitelistCtx(state)
+				if err != nil {
+					return res, errors.Wrap(err, "throttle: context creation")
+				}
+				contractToInfoMap, err := udw.GetContractInfoMapping(ctx)
+				if err != nil {
+					return res, errors.Wrap(err, "throttle: contractToTierMap creation")
+				}
+				TxLimiter.contractToInfoMap = contractToInfoMap
+				// TxLimiter.lastUpdated = time.Now().Unix()
+			}
+			contractAddr := loom.UnmarshalAddressPB(msg.To)
+			//check if contract in list
+			contractInfo, ok := TxLimiter.contractToInfoMap[contractAddr.String()]
+			if !ok {
+				return loomchain.TxHandlerResult{}, ErrContractNotWhitelisted
+			}
+			// if its tierInfo is empty and its expiry is reached, update tier info
+			if contractInfo.TierInfo.BlockRange == 0 ||
```

Comment:
Keep tierID -> tierInfo mapping in different data structure, otherwise it will be lots of duplicate data.

Reply:
Also, I feel keeping expiry time for every contract is a bit overhead. What we need is to have expiry time for tierID -> tierInfo structure. So, that if we change tier information, we could reload it automatically.